src/desktop/WebDialog.ts

```diff
@@ -0,0 +1,149 @@
+	) {
+	}
+
+	makeRequest<T>(requestSender: (facade: FacadeType) => Promise<T>): Promise<T> {
+		return Promise
+			.race([
+				this.closedPromise,
+				requestSender(this.facade)
+			])
+			.catch((e) => {
+				console.log("web dialog error!", e)
+				throw e
+			})
+			.finally(() => {
+				if (!this.browserWindow.isDestroyed()) this.browserWindow.close()
+			})
+	}
+
+	cancel() {
+		this.browserWindow.close()
+	}
+}
+
+export class WebDialogController implements IWebDialogController {
+	constructor(
+		private readonly ipcHandler: WebDialogIpcHandler
+	) {
+	}
+
+	async create<FacadeType extends object>(parentWindowId: number, urlToOpen: URL): Promise<WebDialog<FacadeType>> {
+		const bw = await this.createBrowserWindow(parentWindowId)
+		const closeDefer = defer<never>()
+		bw.on("closed", () => {
+			console.log("web dialog window closed")
+			closeDefer.reject(new CancelledError("Window closed"))
+		})
+
+		register(bw, "F12", () => {
+			bw.webContents.openDevTools()
+		})
+
+		bw.once('ready-to-show', () => bw.show())
+		bw.webContents.on("did-fail-load", () => closeDefer.reject(new Error(`Could not load web dialog at ${urlToOpen}`)))
+		await bw.loadURL(urlToOpen.toString())
+
+		const facade = await this.initRemoteWebauthn<FacadeType>(bw.webContents)
+
+		bw.webContents.on("destroyed", () => this.uninitRemoteWebauthn(bw.webContents))
+
+		return new WebDialog(facade, closeDefer.promise, bw)
+	}
+
+	private async createBrowserWindow(parentWindowId: number) {
+		const active = BrowserWindow.fromId(parentWindowId)
+
+		return new BrowserWindow({
+			parent: active ?? undefined,
+			modal: true,
+			skipTaskbar: true,
+			resizable: false,
+			movable: false,
+			alwaysOnTop: true,
+			fullscreenable: false,
+			frame: false,
+			width: 400,
+			height: 200,
+			autoHideMenuBar: true,
+			show: false,
+			webPreferences: {
+				nodeIntegration: false,
+				nodeIntegrationInWorker: false,
+				nodeIntegrationInSubFrames: false,
+				sandbox: true,
+				contextIsolation: true,
+				webSecurity: true,
+				// @ts-ignore see: https://github.com/electron/electron/issues/30789
+				enableRemoteModule: false,
+				allowRunningInsecureContent: false,
+				preload: path.join(app.getAppPath(), "./desktop/preload-webdialog.js"),
+				webgl: false,
+				plugins: false,
+				experimentalFeatures: false,
+				webviewTag: false,
+				disableDialogs: true,
+				navigateOnDragDrop: false,
+				autoplayPolicy: "user-gesture-required",
+				enableWebSQL: false,
+				spellcheck: false,
+			},
+		})
+	}
+
+	async initRemoteWebauthn<FacadeType>(webContents: WebContents): Promise<FacadeType> {
+		const deferred = defer<void>()
+		const transport = new ElectronWebContentsTransport<WebDialogIpcConfig, "facade", "init">(webContents, this.ipcHandler)
+		const dispatcher = new MessageDispatcher<NativeToWebRequest, WebToNativeRequest>(transport, {
+			"init": () => {
+				deferred.resolve()
+				return Promise.resolve()
+			}
+		})
+		await deferred.promise
```

Comment:
won't this get stuck? I can't see how `"init"` can happen after constructing the transport and dispatcher

Reply:
A have a feeling it is handled somewhere but I might be mixing it up

Reply:
nativeTransport.init gets called from the webView as soon as it creates the NativeWebauthnView and tells us that the communication channel is set up. Or am I misunderstanding what the problem is?

Reply:
that sounds right. It's just very hard to tell when looking at the code. Would be worth a comment